libraries/AP_Mission/AP_Mission.cpp

```diff
@@ -2572,6 +2580,18 @@ bool AP_Mission::calc_rewind_pos(Mission_Command& rewind_cmd)
     return true;
 }
 
+//get first navigation command which also has valid location
+bool AP_Mission::get_first_nav_cmd_with_loc(Mission_Command& cmd) {
+    for (uint8_t i = 1; i < (unsigned)_cmd_total; i++) {
+        if (get_next_nav_cmd(i, cmd)) {
```

Comment:
This is a little odd, your scanning the whole mission, but using `get_next_nav_cmd`. So if you have two none nave commands followed by nav command you would actually check the nav command three times. 

I think is better to just use `get_next_cmd` and add add the `is_nav_cmd` check to the if on 2587

Reply:
We should also start from the current mission item rather than the first index. That would catch the case where the user has forgotten to reset the mission after doing some of it previously.

Reply:
not sure how that works? user flys some part of a mission, switches modes, lands...then, while still powered AND armed, drives 4 miles away with it and tries to fly again in AUTO while still powered?  as I understand it, any disarm/arm (except while still in AUTO mode) restarts the mission at the first waypoint.....works that way on plane at least

Reply:
It depends on the `MIS_RESTART` param, its 0 by default to continue from last point. But your right we should deal with both cases, there is some logic for this here:
https://github.com/ArduPilot/ardupilot/blob/7683b36e3b41306677714b9e1cebb087dead7af3/libraries/AP_Mission/AP_Mission.cpp#L183-L186

Reply:
I think you misunderstand....in Plane and Copter (verified in SITL) if you start a mission, then change modes and land, disarm, and then re-arm...the mission cmd index is ALWAYS reset to the beginning of the mission...independent of any MIS parameter....so you only need ever check the first mission items distance....unless you start a mission, then change mode and fly a long way away from the mission, and then change back to AUTO WITH MIS_RESTART=0 (default) to continue the mission....I dont think you need to check distance in that case, since the user is aware of his situation and has control of his intent....once disarmed and moved to another location, the mission cmd index is reset to the beginning and the first NAV Waypoint check would work...

Reply:
The index is reset on disarming, all the time on copter. Just if not in auto on plane. So you could manually set the wp index before arming again and not start at the beginning. I agree that its unlikely, but it can happen. It only costs us a few lines to check from the current index, I don't see the disadvantage.

Reply:
the disadvantage would be if you set the first waypoint close to prevent the problem this addresses...then pause the mission, the next waypoint upon resume could be much farther than wp1 from home...and you would be denied mode change....